### ISSUE TITLE
OpenTelemetry not setting up correctly

### ISSUE THREAD

**Issue by petrroda** [quarkusio/quarkus#37718]
### Describe the bug

When starting we get the following warnings and no traces are produced

2023-12-13 14:59:57,546 WARN  [io.qua.ope.run.tra.int.ver.OpenTelemetryVertxTracingFactory$VertxDelegator] (vert.x-eventloop-thread-1) VertxTracer delegate not set. Will not submit this trace. SpanKind: RPC; Request: io.vertx.core.http.impl.HttpRequestHead@4ae4ef8d; Operation:GET.
2023-12-13 14:59:57,664 WARN  [io.qua.ope.run.tra.int.ver.OpenTelemetryVertxTracingFactory$VertxDelegator] (vert.x-eventloop-thread-1) VertxTracer delegate not set. Will not submit this trace. Response: io.vertx.core.http.impl.HttpResponseHead@69db63d3; Failure: null.

2023-12-13 14:59:57,694 WARN  [io.qua.ope.run.tra.int.ver.OpenTelemetryVertxTracingFactory$VertxDelegator] (vert.x-eventloop-thread-1) VertxTracer delegate not set. Will not submit this trace. SpanKind: RPC; Request: io.vertx.core.http.impl.HttpRequestHead@c05b10; Operation:GET.

2023-12-13 14:59:57,809 WARN  [io.qua.ope.run.tra.int.ver.OpenTelemetryVertxTracingFactory$VertxDelegator] (vert.x-eventloop-thread-1) VertxTracer delegate not set. Will not submit this trace. Response: io.vertx.core.http.impl.HttpResponseHead@37ba70bb; Failure: null.

**Comment by brunobat** [comment]
This happens because of a known startup race condition. The only way to fix this is to store these traces and send them after OTel is instantiated, however the timestamps will be a mess.

**Comment by goldmann** [comment]
Ok, I have a question then -- does it mean that we kind of loose a few traces at the boot time because of this, but later, when the OTEL subsystem is running -- traces will be published as expected?

**Comment by brunobat** [comment]
No. Right now the traces in the log are lost.
Please mind this is an edge case. Usually apps don't start receiving load until the healthcheck is up which gives plenty of time for everything to start. 
Something is eagerly calling an endpoint when the app is still starting.

**Comment by goldmann** [comment]
I have a feeling you said the same thing as me :D What I meant in the second half of the sentence is that **other** traces will be published normally. At least this is what my collector says.

Either case, these warnings are pretty much the first thing in the log, at least from what I can see:

```
INFO exec -a "java" java -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager -cp "." -jar /deployments/quarkus-run.jar 
INFO running in /deployments
Picked up JAVA_TOOL_OPTIONS: -XX:MaxRAMPercentage=70.0 -XX:+ExitOnOutOfMemoryError
 .d8888b.  888888b.    .d88888b.  888b     d888                  
d88P  Y88b 888  "88b  d88P" "Y88b 8888b   d8888                  
Y88b.      888  .88P  888     888 88888b.d88888                  
 "Y888b.   8888888K.  888     888 888Y88888P888  .d88b.  888d888 
    "Y88b. 888  "Y88b 888     888 888 Y888P 888 d8P  Y8b 888P"   
      "888 888    888 888     888 888  Y8P  888 88888888 888     
Y88b  d88P 888   d88P Y88b. .d88P 888   "   888 Y8b.     888     
 "Y8888P"  8888888P"   "Y88888P"  888       888  "Y8888  888     
                                                                 
                                         Powered by Quarkus 3.7.2
08:47:29,820 INFO  traceId=, parentId=, spanId=, sampled= [io.qua.oid.com.run.OidcCommonUtils] (main) Connecting to OpenId Connect Provider for up to 15 times every 2 seconds mdc:[{}]
08:47:31,818 WARN  traceId=, parentId=, spanId=, sampled= [io.qua.ope.run.tra.int.ver.OpenTelemetryVertxTracingFactory$VertxDelegator] (vert.x-eventloop-thread-1) VertxTracer delegate not set. Will not submit this trace. SpanKind: RPC; Request: io.vertx.core.http.impl.HttpRequestHead@1d185e90; Operation:GET. mdc:[{}]
08:47:31,919 WARN  traceId=, parentId=, spanId=, sampled= [io.qua.ope.run.tra.int.ver.OpenTelemetryVertxTracingFactory$VertxDelegator] (vert.x-eventloop-thread-1) VertxTracer delegate not set. Will not submit this trace. Response: io.vertx.core.http.impl.HttpResponseHead@3b7d824f; Failure: null. mdc:[{}]
08:47:32,035 WARN  traceId=, parentId=, spanId=, sampled= [io.qua.ope.run.tra.int.ver.OpenTelemetryVertxTracingFactory$VertxDelegator] (vert.x-eventloop-thread-1) VertxTracer delegate not set. Will not submit this trace. SpanKind: RPC; Request: io.vertx.core.http.impl.HttpRequestHead@60dfc063; Operation:GET. mdc:[{}]
08:47:32,117 WARN  traceId=, parentId=, spanId=, sampled= [io.qua.ope.run.tra.int.ver.OpenTelemetryVertxTracingFactory$VertxDelegator] (vert.x-eventloop-thread-1) VertxTracer delegate not set. Will not submit this trace. Response: io.vertx.core.http.impl.HttpResponseHead@6a86bd1f; Failure: null. mdc:[{}]
08:47:36,525 INFO  traceId=, parentId=, spanId=, sampled= [io.qua.sch.run.SimpleScheduler] (main) No scheduled business methods found - Simple scheduler will not be started mdc:[{}]
08:47:36,631 INFO  traceId=, parentId=, spanId=, sampled= [org.jbo.sbo.ser.fea.sbo.fea.umb.AmqpUmbClientOptionProducer] (main) Setting up AMQP client options mdc:[{}]
...
```

I'm using Quarkus Operator SDK, I think it may be related to the constant GET/PATCH requests it sends to the cluster. But on other hand, I don't think it is initialized even at this stage. I'll try to get more information about it.

**Comment by brunobat** [comment]
Looks like this is sending requests right away after it's instantiated: OidcCommonUtils

**Comment by goldmann** [comment]
Yah, I can confirm it. I run it with DEBUG and clearly the OIDC extension tries to connect to the specified auth server just before the warning appear.

Dunno, I'm obviously not an expert here, but maybe the solution could be to adjust the order of these subsystems initialization and maybe the correct approach would be to make OTEL initialized earlier?

**Comment by brunobat** [comment]
@goldmann, unfortunately I haven't devise a way to start OTel before Vert.x but the OTel exporter actually needs Vert.x, therefore, this is will never be solved based on who starts first.
The only solution I envision is to cache spans and replay them later, however, time stamps might not be correct. If you want to create a new feature request for this caching, I'm ok with it.
What you are doing is also a bit risky because you are issuing a request before Quarkus is completely up. Depending on how you handle the response, you might get into trouble. 
If you really need those spans, I recommend you delay issuing of that 1st request.

**Comment by goldmann** [comment]
I'm not that interested in these spans, no. What I would like is to get rid of this warning.

From what can I see, I don't control when the call is made, it's in control of the OIDC extension. I just configure it. So maybe in my case it would be the OIDC extension what would need to me hold for a bit until other subsystems are ready?

IMHO replying spans is not the way to go.

**Comment by brunobat** [comment]
Do you think we should downgrade the the message to INFO instead of WARN?

**Comment by goldmann** [comment]
I think WARN still makes sense here. It clearly states that something that supposed to be sent isn't. Sometimes it may be important to receive these events. I wouldn't change it.

**Comment by brunobat** [comment]
If you don't mind I'll close this issue as a won't fix. No action will be taken.

**Comment by goldmann** [comment]
I was actually hoping that some of the OIDC extension maintainers could chime in to hear what they think about it. @sberyozkin maybe?

**Comment by brunobat** [comment]
It might be the case that the OIDC extension is issuing a request too soon, yes, but that would be a bug for OIDC.

**Comment by goldmann** [comment]
Yes, exactly. I obviously can create new ticket, but since this one is exactly about this...

**Comment by brunobat** [comment]
@goldmann please rephrase the description and the title to point to OIDC, I'll reopen the issue then.

**Comment by goldmann** [comment]
Sadly I cannot do it, because it's not my ticket.

**Comment by brunobat** [comment]
Ok, can you please create a new one pointing to this one?
